docs/src/adr/011adr-smt-arrays.md

```diff
@@ -169,11 +169,112 @@ The following changes will be made to implement the new encoding of sets:
+(declare-const tuple_with_2 Tuple_Int) ; <2>
+(declare-const tuple_with_3 Tuple_Int) ; <3>
+
+(declare-const finSucc_domain_0 (Array Tuple_Int Bool))
+(declare-const finSucc_domain_1 (Array Tuple_Int Bool))
+(declare-const finSucc_domain_2 (Array Tuple_Int Bool))
+(declare-const finSucc_domain_3 (Array Tuple_Int Bool))
+
+(assert (= finSucc_domain_0 ((as const (Array Tuple_Int Bool)) false)))  ; {}
+(assert (= finSucc_domain_1 (store finSucc_domain_0 tuple_with_1 true))) ; {<1>}
+(assert (= finSucc_domain_2 (store finSucc_domain_1 tuple_with_2 true))) ; {<1>,<2>}
+(assert (= finSucc_domain_3 (store finSucc_domain_2 tuple_with_3 true))) ; {<1>,<2>,<3>}
+```
+
+The array storing the function's domain is used to guard the definition of the array storing the
+function's mappings, since mappings should only be present for values in the domain. The array
+storing the mappings of `finSucc` is shown below.
+
+```
+(declare-const finSucc_0 (Array Tuple_Int Int))
+(declare-const finSucc_1 (Array Tuple_Int Int))
+(declare-const finSucc_2 (Array Tuple_Int Int))
+(declare-const finSucc_3 (Array Tuple_Int Int))
+
+(assert (ite (select finSucc_domain_3 tuple_with_1)
+             (= finSucc_1 (store finSucc_0 tuple_with_1 2))
+             (= finSucc_1 finSucc_0)))
+(assert (ite (select finSucc_domain_3 tuple_with_2)
+             (= finSucc_2 (store finSucc_1 tuple_with_2 3))
+             (= finSucc_2 finSucc_1)))
+(assert (ite (select finSucc_domain_3 tuple_with_3)
+             (= finSucc_3 (store finSucc_2 tuple_with_3 4))
+             (= finSucc_3 finSucc_2)))
+```
+
+Note that, unlike with the new encoding for sets, we do not use constant arrays. The reason is that
+the function's domain cannot be altered, so the array has to constrain only the values in said domain.
+Function application can be done by simply accessing the array at the index of the passed arguments.
+A function application with arguments outside the function's domain leads to an unspecified result in
+TLA+, which is perfectly captured by unconstrained entries in the SMT array. Below are some examples
+of function application.
+
+```
+(assert (= (select finSucc_3 tuple_with_1) 2)) ; SAT
+(assert (= (select finSucc_3 tuple_with_2) 3)) ; SAT
+(assert (= (select finSucc_3 tuple_with_3) 4)) ; SAT
+
+(declare-const tuple_with_4 Tuple_Int) ; <4>
+(assert (= (select finSucc_3 tuple_with_4) 16)) ; SAT
+```
+
+Although a function's domain cannot be altered, its image can be changed via the TLA+ function
+update operator. The update will be encoded as a guarded array update, as illustrated below;
+attempting to update an entry outside the function's domain will lead to no change happening.
+
+```
+(declare-const finSucc_4 (Array Tuple_Int Int))
+(declare-const finSucc_5 (Array Tuple_Int Int))
+
+(assert (ite (select finSucc_domain_3 tuple_with_1) ; [finSucc EXCEPT ![1] = 9]
+             (= finSucc_4 (store finSucc_3 tuple_with_1 9))
+             (= finSucc_4 finSucc_3)))
```

Comment:
This also produces a linear number of constraints, in contrast to the linear number of constraints in the OOPSLA encoding.